lib/github/oauth.js

```diff
@@ -23,30 +23,21 @@ module.exports = (robot) => {
   });
 
   robot.route().get('/github/oauth/callback', async (req, res) => {
-    // @todo: handle condition where state is expired or has been tampered with
-    const state = await SignedParams.load(req.query.state);
-
-    let command;
-    let teamSlackId;
-    let channelSlackId;
-    let userSlackId;
-    if (state.trigger_id) {
-      command = await PendingCommand.find(state.trigger_id);
-
-      if (!command) {
-        // FIXME: handle issue with expired command
-        throw new Error();
+    let state;
+    try {
+      state = await SignedParams.load(req.query.state);
+    } catch (err) {
+      if (err.name && err.name === 'JsonWebTokenError') {
+        return res.status(400).send(`Error: ${err.message}`);
+      } else if (err.name && err.name === 'TokenExpiredError') {
+        return res.status(400).send(`Error: ${err.message}`);
       }
-      teamSlackId = command.team_id;
-      channelSlackId = command.channel_id;
-      userSlackId = command.user_id;
-    } else if (state.slackAction) {
-      teamSlackId = state.slackAction.team.id;
-      userSlackId = state.slackAction.user.id;
-      channelSlackId = state.slackAction.channel.id;
+      throw err;
```

Comment:
This is the only line that's not covered now and I'm not sure if it should or even can be covered in a clean way? Having to manually ensure that we're only `catch`-ing the errors we care about is a bit of JS-language flaw.

Reply:
What if you do all the error handling for known errors in the `SignedParams.load()` function and you return an object like `{ err, state }` and then here just do

```
try {
  const { err, state } = await SignedParams.load(req.query.state);
  if (err) res.status(400).send(`Error: ${err.message}`); // handled error
} catch (err) {
  throw err; // unhandled error
}
```

Then you can unit test `SignedParams.load` for `JsonWebTokenError` and `TokenExpiredError` errors.

Reply:
In this case, wouldn't the `  throw err; // unhandled error` still be uncovered? 😄 

For what it's worth the `JsonWebTokenError` and `TokenExpiredError` errors _are_ already covered now: https://github.com/integrations/slack/pull/816/files#diff-07e1c12a9232d0340549c220b7392130R130-R183